### ISSUE TITLE
clang-tidy bugprone-branch-clone false-positive with missing include

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#67662]
Consider:

`$ cat /tmp/a.cc`

```c++
template <typename T>
struct S {
  std::unique_ptr<T[]> contents_;

  void f(int capacity) {
   if (capacity) {
      contents_.reset(new T[capacity]);
    } else {
      contents_.reset(nullptr);
    }
  }
};
```

```console
$ build/bin/clang-tidy /tmp/a.cc -checks=bugprone-branch-clone --
1 warning and 1 error generated.
Error while processing /tmp/a.cc.
/tmp/a.cc:3:3: error: use of undeclared identifier 'std' [clang-diagnostic-error]
    3 |   std::unique_ptr<T[]> contents_;
      |   ^
/tmp/a.cc:6:4: warning: if with identical then and else branches [bugprone-branch-clone]
    6 |    if (capacity) {
      |    ^
/tmp/a.cc:8:7: note: else branch starts here
    8 |     } else {
      |       ^
Found compiler error(s).
```

The "if with identical then and else branches [bugprone-branch-clone]" is obviously not correct.

This seems to be an effect of `std::unique_ptr` not being declared. If I add `include <memory>` the warning goes away.

**Comment by PiotrZSL** [comment]
To be honest no issue here, you got error in your code.
Due to error AST is mailformed, and "int" is used instead of std::unique_ptr.

**Comment by zmodem** [comment]
I disagree. The warning is incorrect, so this is a bug. In this case it would be better if clang-tidy bailed out.

**Comment by PiotrZSL** [comment]
This is a common problem that impact also other checks. In theory `clang::Expr::containsErrors()`  and `clang::Decl::isInvalidDecl()` could be used to detect these issues, but I didn't test if they will work, and after that still some error should be emited to user.

**Comment by vbvictor** [comment]
> I disagree. The warning is incorrect, so this is a bug. In this case it would be better if clang-tidy bailed out.

Will be fixed in https://github.com/llvm/llvm-project/issues/196733, there are other (potentially all) checks that don't work with broken AST. I think we should never run analysis on broken code in the first place.
I hope to have time to push that in 24 release so tidy will bail out on broken code

**Comment by zmodem** [comment]
Thank you, that sounds great. But can we keep this open until it's actually fixed?

**Comment by vbvictor** [comment]
> Thank you, that sounds great. But can we keep this open until it's actually fixed?

Sure, we can keep this open. We usually close duplicate issues to reduce churn but single exceptions don't hurn
